### ISSUE TITLE
Revise functions `PrepareLeads`, `PrepareLags`, and `GetFirstDifferences`

### ISSUE THREAD

**Issue by santiagohermo** [JMSLab/eventstudyr#31]
In this issue I will review and improve the code of the functions `PrepareLeads`, `PrepareLags`, and `GetFirstDifferences`.

Things to consider:

- From the [Roadmap](https://github.com/JMSLab/eventstudyr/wiki/Roadmap/fd38fa36653182684b1c40067d43331a3dfc8424): _Make `GetFirstDifferences()`, `PrepareLags()`, and `PrepareLeads()` robust to missing values of timevar. See [here](https://github.com/JMSLab/eventstudyr/pull/24#discussion_r1135987509)._
- From [#28](https://github.com/JMSLab/eventstudyr/issues/28): _Potentially [merge](https://r-pkgs.org/man.html#sec-man-multiple-functions) the documentation for `PrepareLeads()` and `PrepareLags()` since they're so similar_
- From [#28](https://github.com/JMSLab/eventstudyr/issues/28): _Revisit the question of whether these functions should be user-facing_

**Comment by santiagohermo** [comment]
I created this issue @nateschor! I plan to tackle it after we submit to CRAN.

Feel free to edit the opening comment if I forgot of any to-dos here!

**Comment by santiagohermo** [comment]
Sounds great @jmshapir!

**Comment by nateschor** [comment]
> @santiagohermo for sequencing I think we can hold work on this issue until the package is accepted at CRAN (i.e., following #33).
> 
> Thanks!

@jmshapir @santiagohermo I found the [CRAN log](https://r-hub.github.io/cransays/articles/dashboard.html) which detials what packages have been submitted, when they were submitted, and what phase they are in. I did some searching for average wait times for CRAN acceptance but I've had a hard time finding estimates.

Does this information in the CRAN log change wanting to hold work on the issue until we're on CRAN? Would you want to work on this issue in parallel with #33 and then we can resubmit to CRAN in a few months after we've made some more improvements to the package?

Thanks!

**Comment by jmshapir** [comment]
@nateschor thanks! I think we can continue to prioritize submitting to CRAN. Once we have submitted there we can discuss whether to start #31 or keep waiting.

**Comment by santiagohermo** [comment]
Update: I started this issue in branch `issue31_leads_lags_diffs`. My first changes were to make the computation of first differences robust to unbalanced panels. Now, whenever the panel is unbalanced we 

1. Create a dataset with all combinations of id and time
2. Compute first differences there, thus adding NAs when either value in the first difference is missing
3. Add back the first-differenced variable to the original dataset

This robustness should carry over to functions that compute leads and lags. I also changed the name of the function from `GetFirstDifferences` to `ComputeFirstDifferences`, as this latter name seems more descriptive.

My next steps will be to merge `PrepareLeads` and `PrepareLags` in a single function. Tentative name is `ComputeShifts`.

**Comment by santiagohermo** [comment]
One more thing: before the function allowed for computing differences without grouping by id. Since we never use this feature in the code, and I don't see use cases it in the context of panel data, I dropped it.

**Comment by santiagohermo** [comment]
Thanks @jmshapir! I didn't know about this functionality, I'll use it next time :)

**Comment by santiagohermo** [comment]
Thought: at the moment we use [the expensive procedure](https://github.com/JMSLab/eventstudyr/issues/31#issuecomment-1522260946) to create FD if the dataset is unbalanced. In reality, we can have an unbalanced dataset but no holes in the time var. 

We need to use the expensive procedure when, for any id, the number of time periods is different than the difference between the last period and the first period. I'll fix this in the end.

**Comment by santiagohermo** [comment]
Thanks @SimonFreyaldenhoven! Interesting to see the evolution of the paper. The subtraction is implemented in the code, I was tempted to remove it but I know we want the code to align with the math so I won't touch it :)

**Comment by santiagohermo** [comment]
I pushed today and finished a first pass of the changes here @jmshapir @nateschor. What I did:

- Added new functions
  - [`ComputeFirstDifferences` ](https://github.com/JMSLab/eventstudyr/blob/f7baf5d368a7cad8ce33e8d156ef96e4a9838d50/R/ComputeFirstDifferences.R) takes the first difference of a variable "robustly". Replaces `GetFirstDifferences`, which was dropped.
  - [`ComputeShifts`](https://github.com/JMSLab/eventstudyr/blob/f7baf5d368a7cad8ce33e8d156ef96e4a9838d50/R/ComputeShifts.R) takes leads/lags of a variable "robustly". Format of shift values is, e.g., `-2, -1, 1`, which is interpreted as `second lead, first lead, first lag`. Replaces `PrepareLeads` and `PrepareLags`, which have been dropped.
- I improved `EventStudy` and `PrepareModelFormula`. First goal was to make them compatible with new functions. Aditionally:
  - Now the coefficients in the output table are ordered (e.g., `z_lead3, z_fd_lead3, .., z_fd, z_fd_lag1, .., z_lag3`)
  - Now when we estimate a static model the coefficient is called simply `z`, before it was called `z_lead0`.
- I updated tests and they seem to work with the changes.

To-dos:
- Implement [this](https://github.com/JMSLab/eventstudyr/issues/31#issuecomment-1531529600)
- Update the branch and update examples with new dataset
- Decide whether the new functions will be user-facing. I think that now they add a bit of value, as they compute leads/lags taking into account potential holes in the time var.

I think we can open a PR as soon as #37 is done

---

Ideas for future improvements:

- Clean up `EventStudyOLS` and `EventStudyFHS`. There is a lot of unnecessary code as inside each `if` we declare the entire estimation. We should go over the ifs first to define the fixed effects, store them in a formula, and then write the code that calls `estimatr` once in the end using that formula.
- Related to [this](https://github.com/JMSLab/eventstudyr/issues/31#issuecomment-1531529600). I think it would be useful to define a new function that checks the panel structure of the data. It would
  1. Check whether panel is unbalanced or not
  2. Check whether there are any holes in time periods within each id
  3. Return two logical values indicating whether these are TRUE or FALSE.
  This function would simplify the code `EventStudy` and could be shipped separately.
- We had a few things on static models in [the Roadmap](https://github.com/JMSLab/eventstudyr/wiki/Roadmap/9260944bdc55414aaf18c9d6a16405191ba937ff). Given that I made some improvements here already, we could open a new issue to double-check what I did fixes our previous concerns and make `EventStudyPlot` work for static models.

**Comment by jmshapir** [comment]
@santiagohermo thank you!

Regarding the ideas for improvements, can you add them to the roadmap (in either short term or someday, depending on how you think we should prioritize)? And can you remove anything from the roadmap that is now implemented (even if only on a branch)?

Thanks!

**Comment by nateschor** [comment]
Per call: We'll take this to PR after wrapping up #37. I won't have much time next week to review so we'll need more help with reviewing the code.

**Comment by nateschor** [comment]
> Per call: We'll take this to PR after wrapping up #37. I won't have much time next week to review so we'll need more help with reviewing the code.

@santiagohermo heads up that we just wrapped up #37!

**Comment by santiagohermo** [comment]
Thanks @nateschor! Branch for #31 was updated from `main`.

One note: It looks like the README figure ended up being missing in the last commit, see [here](https://github.com/JMSLab/eventstudyr/tree/fb7edc815c626c138b2d25f4cd8271411d15b91e#usage). I added it back in https://github.com/JMSLab/eventstudyr/commit/15a3760121b192b22aae02254542479c00c517e8 in this branch.

**Comment by nateschor** [comment]
I'm sorry about that, thanks for pointing it out @santiagohermo! I was getting a proxy related error when I tried knitting the document and assumed that it wouldn't cause any changes, but it clearly did.

How can we fix it? The two solutions I'm thinking of to update our `README` on `main` are:
* pull the version from this issue branch
* open a pull on a separate branch just for fixing the `README`

What do you think?

Thanks!

**Comment by santiagohermo** [comment]
Thanks @nateschor! If you don't mind waiting for a few days, then once we merge this branch the fix will be implemented. If you are in a rush I can just knit the README in master and add the figure there (like I did in https://github.com/JMSLab/eventstudyr/commit/15a3760121b192b22aae02254542479c00c517e8).

**Comment by jmshapir** [comment]
FWIW I think it's fine to have the readme as is in `main` and wait until we wrap up #31 to repair it. There is still a live link to the vignette for those who want to see an example plot.

**Comment by santiagohermo** [comment]
Update:

- I finished a few to dos [here](https://github.com/JMSLab/eventstudyr/issues/31#issuecomment-1532066993), it remains to discuss whether the new functions should be user-facing or not.
- [I changed the Roadmap](https://github.com/JMSLab/eventstudyr/wiki/Roadmap/_compare/9260944bdc55414aaf18c9d6a16405191ba937ff...d7711a0321f2274073b7bcef35dce6def62c613f) to drop some things that were already implemented and add the ideas I mention [here](https://github.com/JMSLab/eventstudyr/issues/31#issuecomment-1532066993).

I'll move to PR.

**Comment by santiagohermo** [comment]
Continues in #38

**Comment by santiagohermo** [comment]
Summary: In this issue we

- Replaced `GetFirstDifferences()` with `ComputeFirstDifferences()`, which computes differences robustly to gaps in the time variable
- Replaced both `PrepareLeads()` and `PrepareLags()` with `ComputeShifts()`, which is also robust to gaps in time variable
- Improved `PrepareModelFormula()` and `EventStudy()` to display estimated coefficients more nicely and handle static models
- Updated relevant tests for these functions

Changes merged to master in https://github.com/JMSLab/eventstudyr/commit/bec662adf1b388aa7769c3b4fe70d2a47eaf55d9
